src/python/pants/backend/jvm/tasks/jvm_compile/javac/javac_compile.py

```diff
@@ -136,12 +138,27 @@ def compile(self, ctx, args, classpath, upstream_analysis,
         settings_args = (a.replace('$JAVA_HOME', distribution.home) for a in settings.args)
       javac_cmd.extend(settings_args)
 
-    javac_cmd.extend([
-      '-d', ctx.classes_dir,
-      # TODO: support -release
-      '-source', str(settings.source_level),
-      '-target', str(settings.target_level),
-    ])
+    if self.execution_strategy() == 'hermetic':
+      javac_cmd.extend([
+        # We need to strip the source root from our output files. Outputting to a directory, and
+        # capturing that directory, does the job.
+        # Unfortunately, javac errors if the directory you pass to -d doesn't exist, and we don't
+        # have a convenient way of making a directory in the output tree, so let's just use the
+        # working directory as our output dir.
+        # This also has the benefit of not needing to strip leading directories from the returned
+        # snapshot.
+        '-d', '.',
+        # TODO: support -release
+        '-source', str(settings.source_level),
```

Comment:
The source and target flags should probably be added unconditionally, rather than in the if block :)